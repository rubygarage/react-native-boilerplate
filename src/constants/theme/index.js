export default {
  colors: {
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
  },
  sizes: {
  },
};
