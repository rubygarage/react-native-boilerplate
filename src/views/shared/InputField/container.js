import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { call, path, pathOr } from 'ramda';
import { injectIntl } from 'react-intl';
import { withTheme } from 'styled-components/native';

import InputFieldComponent from './component';

class InputField extends Component {
  state = {
    isFocused: false,
    actionWidth: 0,
  }

  onActionRendered = (event) => {
    const { width } = event.nativeEvent.layout;
    this.setState({ actionWidth: width });
  }

  onFocus = () => {
    this.setState({ isFocused: true });
  }

  onBlur = (event) => {
    const { field: { name }, form: { handleBlur } } = this.props;
    call(handleBlur(name), event);
    this.setState({ isFocused: false });
  }

  getErrorData = () => {
    const { field: { name }, form: { errors } } = this.props;
    const errorId = path([name, 'id'], errors);
    const errorValues = path([name, 'values'], errors);
    const errorIcon = pathOr('alert', [name, 'icon'], errors);
    return { errorId, errorValues, errorIcon };
  }

  render() {
    const { actionWidth, isFocused } = this.state;
    return (
      <InputFieldComponent
        {...this.props}
        onActionRendered={this.onActionRendered}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
        actionWidth={actionWidth}
        isFocused={isFocused}
        {...this.getErrorData()}
      />
    );
  }
}

InputField.propTypes = {
  field: PropTypes.shape({ name: PropTypes.string.isRequired }).isRequired,
  form: PropTypes.shape({
    errors: PropTypes.shape().isRequired,
    handleBlur: PropTypes.func.isRequired,
  }).isRequired,
  theme: PropTypes.shape({ colors: PropTypes.shape.isRequired }).isRequired,
};

const enhance = compose(
  injectIntl,
  withTheme,
);

export { InputField as InputFieldContainer };
export default enhance(InputField);
