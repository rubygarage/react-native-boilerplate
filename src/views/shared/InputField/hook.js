import { useState } from 'react';
import { call, path, pathOr } from 'ramda';
import { useIntl } from 'react-intl';

import useTheme from 'lib/hooks/useTheme';

function useContainer(props) {
  const {
    field: { name },
    form: { handleBlur, errors },
  } = props;

  const intl = useIntl();
  const theme = useTheme();

  const [isFocused, setIsFocused] = useState(false);
  const [actionWidth, setActionWidth] = useState(0);

  /**
   * On Action required.
   */
  const onActionRendered = (event) => {
    const { width } = event.nativeEvent.layout;

    setActionWidth(width);
  };

  /**
   * On focus handler.
   */
  const onFocus = () => {
    setIsFocused(true);
  };

  /**
   * On Blur handler
   */
  const onBlur = (event) => {
    call(handleBlur(name), event);
    setIsFocused(false);
  };

  /**
   * Error data
   */
  const getErrorData = () => {
    const errorId = path([name, 'id'], errors);
    const errorValues = path([name, 'values'], errors);
    const errorIcon = pathOr('alert', [name, 'icon'], errors);
    return { errorId, errorValues, errorIcon };
  };

  return {
    isFocused,
    actionWidth,
    onActionRendered,
    onFocus,
    onBlur,
    getErrorData,
    setIsFocused,
    setActionWidth,
    intl,
    theme,
  };
}

export default useContainer;
