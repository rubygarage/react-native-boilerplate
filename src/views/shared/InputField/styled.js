import styled from 'styled-components/native';

const Wrap = styled.View`
  flex-direction: column;
`;

const Label = styled.Text`
  margin-bottom: 4;
  font-size: 14;
  line-height: 20;
  color: ${({ theme }) => theme.colors.black};
`;

const FieldWrap = styled.View`
  position: relative;
`;

const Field = styled.TextInput`
  border-radius: 3;
  min-height: 48;
  border-width: ${({ isFocused, isError }) => (isFocused && !isError ? 2 : 1)};
  border-style: solid;
  border-color: ${({ theme, isError, isFocused }) => {
    if (isError) {
      return theme.colors.black;
    }
    if (isFocused) {
      return theme.colors.black;
    }
    return theme.colors.black;
  }};
  padding-top: 10;
  padding-right: ${({ actionWidth }) => (actionWidth + 12)};
  padding-bottom: 10;
  padding-left: 12;
  font-size: 14;
  line-height: 20;
  color: ${({ theme }) => theme.colors.black};
`;

const FieldError = styled.View`
  flex-direction: row;
  align-items: center;
  margin-top: 6;
  margin-bottom: 8;
`;

const FieldErrorClose = styled.View`
  margin-right: 8;
`;

const FieldErrorMessage = styled.Text`
  color: ${({ theme }) => theme.colors.black};
  font-size: 14;
  line-height: 20;
`;

const ActionButtonWrapper = styled.View`
  position: absolute;
  height: 48;
  right: 8;
  justify-content: center;
`;

export {
  Wrap, Label, FieldWrap, Field,
  FieldError, FieldErrorClose, FieldErrorMessage, ActionButtonWrapper,
};
