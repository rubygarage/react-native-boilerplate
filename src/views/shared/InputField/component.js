import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import Icon from 'views/shared/Icon';
import Button from 'views/shared/Button';
import * as S from './styled';

const InputFieldComponent = ({
  labelId,
  placeholderId,
  errorId,
  errorValues,
  errorIcon,
  actionId,
  theme,
  onActionRendered,
  onFocus,
  onBlur,
  actionWidth,
  onActionClick,
  isFocused,
  field: { name, value },
  form: { handleChange },
  intl: { formatMessage },
}) => (
  <S.Wrap>
    {labelId
      && (
        <S.Label>
          <FormattedMessage id={labelId} />
        </S.Label>
      )}
    <S.FieldWrap>
      <S.Field
        placeholder={formatMessage({ id: placeholderId })}
        onChangeText={handleChange(name)}
        onBlur={onBlur}
        onFocus={onFocus}
        value={value}
        actionWidth={actionWidth}
        isFocused={isFocused}
        isError={Boolean(errorId)}
      />
      {errorId
        && (
          <S.FieldError>
            <S.FieldErrorClose>
              <Icon name={errorIcon} size={16} color={theme.colors.red500} />
            </S.FieldErrorClose>
            <S.FieldErrorMessage>
              <FormattedMessage id={errorId} values={errorValues} />
            </S.FieldErrorMessage>
          </S.FieldError>
        )}
      {onActionClick && actionId
        && (
          <S.ActionButtonWrapper onLayout={onActionRendered}>
            <Button isSmallWidth isSmallHeight onPress={onActionClick}>
              <FormattedMessage id={actionId} />
            </Button>
          </S.ActionButtonWrapper>
        )}
    </S.FieldWrap>
  </S.Wrap>
);

InputFieldComponent.propTypes = {
  placeholderId: PropTypes.string.isRequired,
  onActionRendered: PropTypes.func.isRequired,
  onFocus: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,
  actionWidth: PropTypes.number.isRequired,
  isFocused: PropTypes.bool.isRequired,
  field: PropTypes.shape({
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
  }).isRequired,
  errorValues: PropTypes.shape(),
  form: PropTypes.shape({ handleChange: PropTypes.func.isRequired }).isRequired,
  intl: PropTypes.shape({ formatMessage: PropTypes.func.isRequired }).isRequired,
  theme: PropTypes.shape({ colors: PropTypes.shape().isRequired }).isRequired,
  onActionClick: PropTypes.func,
  labelId: PropTypes.string,
  errorId: PropTypes.string,
  errorIcon: PropTypes.string,
  actionId: PropTypes.string,
};

InputFieldComponent.defaultProps = {
  onActionClick: undefined,
  labelId: undefined,
  errorId: undefined,
  errorIcon: undefined,
  actionId: undefined,
  errorValues: undefined,
};

export default InputFieldComponent;
