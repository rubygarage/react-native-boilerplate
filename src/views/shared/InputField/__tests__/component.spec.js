import React from 'react';
import { shallow } from 'enzyme';

import theme from 'utils/testHelpers/mockedTheme';
import testIntl from 'utils/testHelpers/testIntl';
import InputField from '../component';

const defaultProps = {
  labelId: 'label_id',
  placeholderId: 'placeholder_id',
  theme,
  onActionRendered: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
  borderColor: 'white',
  borderWidth: 2,
  actionWidth: 0,
  form: { handleChange: jest.fn() },
  field: { name: 'testName', value: 'value' },
  isFocused: false,
  intl: testIntl,
};

describe('InputField component', () => {
  const component = shallow(<InputField {...defaultProps} />);

  it('renders correctly default', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders without label', () => {
    const props = {
      ...defaultProps,
      labelId: undefined,
    };
    component.setProps(props);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with error', () => {
    const props = {
      ...defaultProps,
      errorId: 'error_id',
      errorIcon: 'close',
    };
    component.setProps(props);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with action', () => {
    const props = {
      ...defaultProps,
      onActionClick: jest.fn(),
      actionId: 'action_id',
    };
    component.setProps(props);
    expect(component).toMatchSnapshot();
  });
});
