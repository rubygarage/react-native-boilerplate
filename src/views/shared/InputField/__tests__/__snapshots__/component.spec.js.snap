// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField component renders correctly default 1`] = `
<Styled(View)>
  <Styled(Text)>
    <FormattedMessage
      id="label_id"
      values={Object {}}
    />
  </Styled(Text)>
  <Styled(View)>
    <Styled(TextInput)
      actionWidth={0}
      isError={false}
      isFocused={false}
      onBlur={[MockFunction]}
      onFocus={[MockFunction]}
      placeholder="{Translation id: placeholder_id}"
      value="value"
    />
  </Styled(View)>
</Styled(View)>
`;

exports[`InputField component renders correctly with action 1`] = `
<Styled(View)>
  <Styled(Text)>
    <FormattedMessage
      id="label_id"
      values={Object {}}
    />
  </Styled(Text)>
  <Styled(View)>
    <Styled(TextInput)
      actionWidth={0}
      isError={true}
      isFocused={false}
      onBlur={[MockFunction]}
      onFocus={[MockFunction]}
      placeholder="{Translation id: placeholder_id}"
      value="value"
    />
    <Styled(View)>
      <Styled(View)>
        <Icon
          allowFontScaling={false}
          color="[mocked red500]"
          name="close"
          size={16}
        />
      </Styled(View)>
      <Styled(Text)>
        <FormattedMessage
          id="error_id"
          values={Object {}}
        />
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      onLayout={[MockFunction]}
    >
      <WithTheme(Button)
        isSmallHeight={true}
        isSmallWidth={true}
        onPress={[MockFunction]}
      >
        <FormattedMessage
          id="action_id"
          values={Object {}}
        />
      </WithTheme(Button)>
    </Styled(View)>
  </Styled(View)>
</Styled(View)>
`;

exports[`InputField component renders correctly with error 1`] = `
<Styled(View)>
  <Styled(Text)>
    <FormattedMessage
      id="label_id"
      values={Object {}}
    />
  </Styled(Text)>
  <Styled(View)>
    <Styled(TextInput)
      actionWidth={0}
      isError={true}
      isFocused={false}
      onBlur={[MockFunction]}
      onFocus={[MockFunction]}
      placeholder="{Translation id: placeholder_id}"
      value="value"
    />
    <Styled(View)>
      <Styled(View)>
        <Icon
          allowFontScaling={false}
          color="[mocked red500]"
          name="close"
          size={16}
        />
      </Styled(View)>
      <Styled(Text)>
        <FormattedMessage
          id="error_id"
          values={Object {}}
        />
      </Styled(Text)>
    </Styled(View)>
  </Styled(View)>
</Styled(View)>
`;

exports[`InputField component renders without label 1`] = `
<Styled(View)>
  <Styled(View)>
    <Styled(TextInput)
      actionWidth={0}
      isError={false}
      isFocused={false}
      onBlur={[MockFunction]}
      onFocus={[MockFunction]}
      placeholder="{Translation id: placeholder_id}"
      value="value"
    />
  </Styled(View)>
</Styled(View)>
`;
